### PR TITLE
[mlir][utils] Add script to verify canonicalizations against Alive2

### PR DESCRIPTION
**Pull request by Hardcode84**
This script takes IR before and after canonicalization, translates it into llvm IR and converts it to format suitable for Alive2 https://alive2.llvm.org/ce/

This is primarily for arith canonicalizations verification, but technically it can be adapted for any dialect translatable to llvm.

Usage `python verify_canon.py canonicalize.mlir -f func1 func2 ...`

Example output: https://alive2.llvm.org/ce/z/KhQs4J

Initial discussion: https://github.com/llvm/llvm-project/pull/91646#pullrequestreview-2049342826

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Thanks for writing this. I came up with a similar `arith` rewrite testing flow but never bothered to package it as a script, seems very useful!

File: mlir/utils/verify-canon/verify_canon.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        files.append(tmp)
+
+    return subprocess.check_output(["llvm-link", "-S"] + [f.name for f in files])
+
+
+if __name__ == "__main__":
+    argv = sys.argv
+    if len(argv) < 2:
+        print(f"usage: {argv[0]} canonicalize.mlir [func1] [func2] ...")
+        exit(0)
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/utils/verify-canon/verify_canon.py:45]: I kinda wanted to keep this script simple and avoid using `argparse` unless we actually need to support non-trivial options

[the remark this answers is not in the document]
Comment (kuhar) [reply mlir/utils/verify-canon/verify_canon.py:45]: I don't think argparse adds complexity here -- it's ~5 lines wire up all the flags that you already support but with better ergonomics

File: mlir/utils/verify-canon/verify_canon.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    funcs = argv[2:]
+
+    orig_ir = Path(file).read_bytes()
+    orig_ir = filter_funcs(orig_ir, funcs)
+
+    to_llvm_args = [
+        "--convert-arith-to-llvm",
+        "--convert-func-to-llvm",
+        "--convert-ub-to-llvm",
+        "--convert-vector-to-llvm",
```
Comment (joker-eph) [inline mlir/utils/verify-canon/verify_canon.py:57]: These should all be handled by a single `--convert-to-llvm`, aren't they?

File: mlir/utils/verify-canon/verify_canon.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Run canonicalization, convert IR to LLVM and convert to format suitable to
+# verification against Alive2 https://alive2.llvm.org/ce/.
```
Comment (gysit) [inline mlir/utils/verify-canon/verify_canon.py:6]: I would probably go for an item list here to make it a bit less dense:

e.g.:
```
This script is a helper to verify canonicalization patterns using Alive2 https://alive2.llvm.org/ce/.
It performs the following steps:
- Filters out the provided test functions.
- Runs the canonicalization pass on the remaining functions.
- Lowers both the original and the canonicalized functions to LLVM IR.
- Prints the canonicalized and the original function side-by-side in a format that can be copied into Alive2 for verification.
Example: `python verify_canon.py canonicalize.mlir func1 func2 func3`
```

File: mlir/utils/verify-canon/verify_canon.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/utils/verify-canon/verify_canon.py]: done

File: mlir/utils/verify-canon/verify_canon.py (hunk 2 above)
Comment (Hardcode84) [reply mlir/utils/verify-canon/verify_canon.py]: done

File: mlir/utils/verify-canon/verify_canon.py (hunk 3 above)
Comment (Hardcode84) [reply mlir/utils/verify-canon/verify_canon.py]: done

File: mlir/utils/verify-canon/verify_canon.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# This script is a helper to verify canonicalization patterns using Alive2
+# https://alive2.llvm.org/ce/.
+# It performs the following steps:
+# - Filters out the provided test functions.
+# - Runs the canonicalization pass on the remaining functions.
+# - Lowers both the original and the canonicalized functions to LLVM IR.
+# - Prints the canonicalized and the original function side-by-side in a format
```
Comment (gysit) [inline mlir/utils/verify-canon/verify_canon.py:11]:
```suggestion
# - Prints the canonicalized and the original functions side-by-side in a format
```
Sorry my bad... This should be plural.

**Review by gysit — APPROVED** [review]
Thanks!

LGTM from my end.

**Review by kuhar — APPROVED** [review]

**Comment by jpienaar** [comment]
Just noticed this one, this is really nice. Hoping it flushes out more cases :)

**Pull request merged**